### PR TITLE
Metric aggregation testing

### LINKED ISSUE #3230: Test all metrics against sklearn (with many input trials)
Hand-chosen values are not enough, we need to test with a large batch of inputs where possible. 
Something in this style, maybe with a fixed seed: 
```python
def test_auroc_versus_sklearn():
    for i in range(100):
        target = torch.randint(0, 2, size=(10, ))
        pred = torch.randint(0, 2, size=(10,))
        score_sk = sklearn.metrics.roc_auc_score(target.numpy(), pred.numpy())  # sklearn
        score_pl = auroc(pred, target)  # Pytorch Lightning 
        assert torch.allclose(torch.tensor(score_pl).float(), torch.tensor(score_sk).float())
```

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

With PR #3245 we changed the way we are doing ddp sync and also introduced aggregation over multiple batches. This PR will add test for all metrics such that we are sure that they work in the following cases:
* Aggregation over multiple devices (aka running in ddp mode)
* Aggregation over multiple equal size batches
* Aggregation over multiple unequal size batches (important because a lot of assumptions that the final metric value can be computed as the mean over individual computed values break in this example)
* Aggregation over multiple devices and multiple batches
* Model integration

Should fix #3230

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SkafteNicki** [comment]
@Borda , @awaelchli , @justusschock would like your guys input on this.
The testing of metrics in ddp mode has long been overdue and after doing this to 7 of them clearly shows that they are broken when we start to aggregate (both multi device and multi batch).
It requires that most metrics divide their computations into pre-ddp and post-ddp (as I hinted at in #2528 and borda pointed to in #3510).
Lukily, we do have the interface for this: pre-ddp goes into `forward` method and post-ddp goes in the static method `compute`.
As stated I have done this to 7 of them and would like your guys input before continuing:

1) Is the test setup too complicated? Espicially should testing be split into their respective files (classification, regression ect) or one large as in this PR?

2) How to keep code duplication low with the functional backend? Right now I have basically implemented a `return_state` argument to the functional version of each metric which returns the metric-state right before ddp sync so only minimal computations needs to happen after ddp sync.

3) How to split the work? I can do many of them but I guess one big PR will be too complicated to review.

**Comment by ananyahjha93** [comment]
@SkafteNicki hey, I am am working on this as well, and I think there are some metrics which wouldn't work in that order. I think it is better to change the order to pre_ddp (input_convert) -> ddp_reduce/ddp_gather -> post_ddp (forward/output_convert). i.e. we need to gather statistics before computing the metrics themselves, rather than computing the metrics on individual GPUs and figuring out how metrics from different batches are combined. It won't just be mean/weighted mean or something along those lines for any specific metric. So, I have been changing this itself for metrics but was breaking a lot of things yesterday.

File: tests/metrics/test_metrics.py [hunk 1]
```diff
@@ -190,7 +190,9 @@ def test_saving_pickable(tmpdir, metric: Metric):
     assert results_before_save == results_after_load
 
 
-def check_call_order():
+def test_correct_call_order():
+    """ Check that hooks are called in the expected order """
+
     class DummyMetric(Metric):
```
Comment (awaelchli) [inline tests/metrics/test_metrics.py:196]: this would be too much to change heere, but I just want to comment that I think the right way to go about checking call order etc. is with the python unittest mocking system. it does not require us to write these artificial classes that track calls, because it's built in. something to look out for in the future I think. Also in other places of Lightning. cc @justusschock @Borda

**Review by awaelchli — COMMENTED** [review]
nice, LGTM on a high level. I'm not a metrics guy, so no comment on correctness :)

File: pytorch_lightning/metrics/classification.py [hunk 2]
```diff
@@ -507,29 +543,15 @@ def __init__(
 
             reduce_group: the process group to reduce metric results from DDP
         """
-        super().__init__(
+        super(TensorMetric, self).__init__(
```
Comment (justusschock) [inline pytorch_lightning/metrics/classification.py:546]: can't we just call fbeta init here with beta=1 and then overwrite the name? should be less code, no?

**Review by justusschock — APPROVED** [review]
LGTM. Just left some comments.

File: tests/metrics/test_metrics.py (hunk 1 above)
Comment (SkafteNicki) [reply tests/metrics/test_metrics.py:196]: Lets do that in a follow up PR (not that I have that much experience with mocking)

File: pytorch_lightning/metrics/classification.py (hunk 2 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/classification.py]: I have changed it to your suggestion 👍

**Review by ananyahjha93 — APPROVED** [review]

File: pytorch_lightning/metrics/functional/regression.py [hunk 3]
```diff
@@ -155,25 +172,30 @@ def psnr(
...
     """
     if data_range is None:
-        data_range = max(target.max() - target.min(), pred.max() - pred.min())
+        data_range = target.max() - target.min()
     else:
         data_range = torch.tensor(float(data_range))
 
+    if return_state:
+        return {'data_range': data_range,
+                'squared_error': F.mse_loss(pred, target, reduction='none').sum(),
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/regression.py:196]: just thinking that it would be sager to include the sum in the name...

File: pytorch_lightning/metrics/metric.py [hunk 4]
```diff
@@ -163,17 +163,27 @@ def aggregate(self, *tensors: torch.Tensor) -> torch.Tensor:
...
+        if len(tensors) == 1:
+            tensors = tensors[0]
+            if isinstance(tensors, Mapping):
+                return {k: _stack_and_agg(tensors[k]) for k in tensors.keys()}
+            if isinstance(tensors, list):
+                return _stack_and_agg(tensors)
+            if isinstance(tensors, tuple):
+                return tensors
+            if isinstance(tensors, torch.Tensor):
+                return _stack_and_agg(tensors)
```
Comment (Borda) [inline pytorch_lightning/metrics/metric.py:176]:
```suggestion
            elif isinstance(tensors, list):
                return _stack_and_agg(tensors)
            elif isinstance(tensors, tuple):
                return tensors
            elif isinstance(tensors, torch.Tensor):
                return _stack_and_agg(tensors)
```

File: pytorch_lightning/metrics/metric.py [hunk 5]
```diff
@@ -163,17 +163,27 @@ def aggregate(self, *tensors: torch.Tensor) -> torch.Tensor:
...
+
+        # multiple tensors (from aggregation over batches)
+        if isinstance(tensors[0], Mapping):
+            return {k: torch.stack([tensor[k] for tensor in tensors]).sum(0) for k in tensors[0].keys()}
+        if isinstance(tensors[0], Sequence):
+            return tuple([torch.stack(tmp).sum(0) for tmp in zip(*tensors)])
+        if isinstance(tensors[0], torch.Tensor):
+            return torch.stack(tensors).sum(0)
```
Comment (Borda) [inline pytorch_lightning/metrics/metric.py:184]:
```suggestion
        elif isinstance(tensors[0], Sequence):
            return tuple([torch.stack(tmp).sum(0) for tmp in zip(*tensors)])
        elif isinstance(tensors[0], torch.Tensor):
            return torch.stack(tensors).sum(0)
```

File: pytorch_lightning/metrics/self_supervised.py [hunk 6]
```diff
@@ -56,6 +57,9 @@ def __init__(
         assert reduction in ('none', 'sum', 'mean')
         self.reduction = reduction
 
+        rank_zero_warn('Please note that Metric `EmbeddingSimilarity` does not '
+                       'support aggregation.')
```
Comment (Borda) [inline pytorch_lightning/metrics/self_supervised.py:61]:
```suggestion
        rank_zero_warn('Please note that Metric `EmbeddingSimilarity` does not support aggregation.')
```

File: tests/base/model_test_epoch_ends.py [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 import torch
 
+from pytorch_lightning import EvalResult
```
Comment (Borda) [inline tests/base/model_test_epoch_ends.py:5]: is this used?

File: tests/metrics/test_aggregation.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+from skimage.metrics import (
+    peak_signal_noise_ratio,
+    structural_similarity
+)
+
+# example structure
+TestCase = namedtuple('example', ['name', 'lightning_metric', 'comparing_metric', 'test_input'])
+
+# setup some standard testcases
+N_samples = 200
```
Comment (Borda) [inline tests/metrics/test_aggregation.py:64]:
```suggestion
NB_SAMPLES = 200
```

File: tests/metrics/test_aggregation.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+             multiclass_and_binary_example),
+    TestCase('confusion matrix without normalize',
+             ConfusionMatrix,
+             confusion_matrix,
+             multiclass_and_binary_example),
+    TestCase('confusion matrix with normalize',
+             partial(ConfusionMatrix, normalize=True),
+             partial(confusion_matrix, normalize='true'),
+             multiclass_and_binary_example),
+    # TestCase('precision recall curve',
```
Comment (Borda) [inline tests/metrics/test_aggregation.py:96]: why they are commented, pls add a note with reason...

File: tests/metrics/test_aggregation.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+             regression_example),
+    # TestCase('structual similarity index measure',
+    #          SSIM,
+    #          structural_similarity,
+    #          regression_example)
+]
+
+
+# Utility test functions
+def idsfn(test):
```
Comment (Borda) [inline tests/metrics/test_aggregation.py:172]:
```suggestion
def _idsfn(test):
```

File: pytorch_lightning/metrics/metric.py (hunk 4 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:176]: @Borda if I change from an `if` to `elif`, codefactor will complain about:
"Unnecessary "elif" after "return"

**Review by awaelchli — APPROVED** [review]
